PHP DB code: clean up the logic, and allow for > 1 readonly replica

davidpanderson opened the pull request:
Also always use mysqli interface

brevilo commented:
> Also always use mysqli interface

Please, use atomic commits so that one can review separate topics separately.

brevilo commented on html/inc/boinc_db.inc:
```diff
@@ -27,119 +27,97 @@ incs();
+        $retval = $instance->init_conn($user, $passwd, $host, $name);
+        if ($retval) {
+            if ($instance->do_query("use $name")) {
+                //error_log("BoincDb::get_aux(): connected to DB $dbnum");
```
Can be removed.

brevilo commented on html/inc/boinc_db.inc:
```diff
@@ -27,119 +27,97 @@ incs();
-                        error_log("BoincDb::get_aux(): Couldn't select database $name on $replica_host (giving up)");
-                        $instance = null;
+                    if ($instance->do_query("use $name")) {
+                        //error_log("BoincDb::get_aux(): connected to replica DB $dbnum");
```
Can be removed.

brevilo commented on html/inc/boinc_db.inc:
```diff
@@ -27,119 +27,97 @@ incs();
+        $user = parse_config($config, '<db_user>');
+        $name = parse_config($config, '<db_name>');
+        $passwd = parse_config($config, '<db_passwd>');
+        // OK if host is null; it will use localhost
```
How? I don't see it being set explicitly. Do you depend on `mysqli` to do that? If so, it [falls back](https://www.php.net/manual/en/mysqli.construct.php) to use [`mysqli.default_host`](https://www.php.net/manual/en/mysqli.configuration.php#ini.mysqli.default-host) which doesn't have a [default value](https://www.php.net/manual/en/mysqli.configuration.php) configured, and I can't find one explicitly defined here either.

brevilo replied:
Thanks

brevilo commented:
Apart from the above I like the approach. Flexible while not overly complex.

davidpanderson commented:
I removed the 'use db_name' query (not needed with mysqli)
and made the default 'localhost' explicit.

brevilo commented on html/inc/boinc_db.inc:
```diff
@@ -57,31 +57,27 @@ class BoincDb extends DbConn {
-                        self::$instance = $instance;
-                        return;
-                    }
+                    //error_log("BoincDb::get_aux(): connected to replica DB $dbnum");
```
Please remove dead code like this. You can add it back for debugging if needed. Alternatively, use pre-processor macros.

brevilo replied:
Fixed by @bema-aei

brevilo commented on html/inc/boinc_db.inc:
```diff
@@ -57,31 +57,27 @@ class BoincDb extends DbConn {
-                self::$instance = $instance;
-                return;
-            }
+            //error_log("BoincDb::get_aux(): connected to DB $dbnum");
```
Please remove dead code like this. You can add it back for debugging if needed. Alternatively, use pre-processor macros.

brevilo replied:
Fixed by @bema-aei

brevilo commented:
@bema-aei should ultimately review this as well.

brevilo commented on html/inc/boinc_db.inc:
```diff
@@ -27,119 +27,93 @@ incs();
-                } else {
-                    error_page("Can't connect to database");
-                }
+        if (isset(self::$instance)) {
```
Hm, this implies that there can only be a single instance at a time. What we need here is a hash of instances (indexed by `$dbnum`) such that multiple instances are retained and/or can be switched between. E.g. the server status page needs to query different databases (here: replicas).

bema-aei replied:
I think a hash of instances might be a bit of an overkill, I think E@H is the only project that uses as much as two. However a requirement for us would be to access different replicas at least in sequence. I did this in d1f99043877552ea48ac2912dc4ec9ac9a1efddd

brevilo replied:
Ok, but please add a comment that states that there's only ever be one current DB instance at a time. IOW, the current instance gets created, reused or replaced by a new one.

bema-aei replied:
Is this a requirement?

brevilo replied:
Well no, but that's how you implemented it, isn't it?

brevilo replied:
Resolved [here](https://github.com/BOINC/boinc/pull/5695#discussion_r1698695506)

bema-aei commented:
This works fro us in principle, with the change I mentioned above. I updated my original branch with your changes, with two important changes:
- including a change that affects the whole web code ("use mysqli everywhere") in another commit with very limited scope and not even mentioning it in the commit message is just asking for trouble. It's difficult to review, difficult to test. Worst of all,  if someone of some project later finds it to cause trouble for him, he is unlikely to ever find the commit that breaks his project. So I split your commit in two (leaving the author intact).
- I added the commit I mentioned above that is required to make this work for us. It allows to address multiple DB hosts in sequence without much effort.
- I also changed the debug output to suit my needs, then reverted it. You can re-add it by reverting the 'Revert' commit if needed.
Please have a look at that branch.

AenBleidd commented:
@bema-aei, @brevilo, are you ok with this PR? Does it need anything else (like testing on your side) or we can move forward and merge it?

AenBleidd commented:
@bema-aei, @brevilo, could you please take a look at this PR and test it?
Thank you in advance.

bema-aei commented:
First of all I like the idea of allowing an arbitrary number of replicas. And thank you for implementing a close() function.

[This code](https://github.com/BOINC/boinc/blob/4b4ba9dbf83384f272f083875cb7f187bc6f6f67/html/inc/boinc_db.inc#L107-L109) in get() is still unchanged, i.e. if $instance is set, get() will return the same instance even if $dbnum is different.

At least this doesn't correspond to what I proposed in my branch. Is the "user" of this code required to close one connection / instance before opening another? If so, it should be noted somewhere, I didn't expect that.

bema-aei commented:
Also too I there is an uncaught exception if the DB connection is refused for some reason. I get:
```
PHP Fatal error:  Uncaught mysqli_sql_exception: Connection refused in /srv/BOINC/live-webcode-test/html/inc/db_conn.inc:37
Stack trace:
#0 /srv/BOINC/live-webcode-test/html/inc/db_conn.inc(37): mysqli->__construct()
#1 /srv/BOINC/live-webcode-test/html/inc/boinc_db.inc(65): DbConn->init_conn()
#2 /srv/BOINC/live-webcode-test/html/inc/boinc_db.inc(117): BoincDb::get_aux()
#3 /srv/BOINC/live-webcode-test/html/ops/eah_server_status.php(681): BoincDb::get()
#4 /srv/BOINC/live-webcode-test/html/ops/eah_server_status.php(1108): seconds_behind_master()
#5 {main}
  thrown in /srv/BOINC/live-webcode-test/html/inc/db_conn.inc on line 37
```

davidpanderson commented:
- I added the check for different dbnum in get()
- Yes, you can only use on DB at a time
- An exception would be thrown if had previously called
  mysqli_report(MYSQLI_REPORT_STRICT);
  presumably because you want to handle exceptions.
  But in any case I added a try/catch.

bema-aei commented:
Thanks, this works for us technically. We do use a recent MariaDB, which has some defaults different from MySQL, hence the exception.

However, for reasons I pointed out already I can't really approve the first commit which combines different features w/o even mentioning in the commit message, so I can't recommend to merge this branch as it is.

brevilo commented:
> We do use a recent MariaDB, which has some defaults different from MySQL, hence the exception.

How does a PHP module's config relate to the MySQL flavor used as the DB backend?

bema-aei commented:
> > We do use a recent MariaDB, which has some defaults different from MySQL, hence the exception.
> 
> How does a PHP module's (`mysqli`) config relate to the MySQL flavor used as the DB backend?

Oh. Sorry, was mislead. Indeed, that's the mysqli config.
